(RE-8707) Add `platform_data`

mwaggett opened the pull request:
This PR adds a `platform_data` hash that maps each platform to where its
artifact and repo_config (if applicable) can be found. This info will get added
to the <ref>.yaml file that gets generated when a package is built, allowing
beaker to consume this info without having to figure out the mapping itself.

puppetcla commented:
CLA signed by all contributors.

mwaggett commented:
puppetserver: http://builds.delivery.puppetlabs.net/puppetserver/5.0.0.master.SNAPSHOT.2017.05.04T1636/artifacts/
puppet-agent: http://builds.delivery.puppetlabs.net/puppet-agent/ac700b4d7e70783701e13079e0c7f3a635f3fe9a/artifacts/

melissa commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 .bundle
 .idea
 Gemfile.lock
+*.swp
```
You shouldn't need this here. When there's a file with this ending, it means you have the file open somewhere.

mwaggett replied:
I know, but I'm often committing while I have another file open and it kept being added to the commit. It was bothering me. But I can take this out if you don't think it belongs in the repo.

melissa replied:
It doesn't belong in the repo

melissa commented on lib/packaging/config/params.rb:
```diff
@@ -125,6 +125,7 @@ module Pkg::Params
                   :pe_platforms,
                   :pe_version,
                   :pg_major_version,
+                  :platform_data,
```
Are you actually using this anywhere? It could be accessed as `Pkg::Config.platform_data`

mwaggett replied:
I think this would be fine without this, but I left it in for consistency with all the other things that end up in that yaml file, even though we're setting it a different way.

melissa replied:
This implies that you'd be able to alter or manipulate this variable in some way outside of the code you've written for it. We're not doing that. It's an unnecessary addition

mwaggett replied:
okay, I'll remove it.

melissa commented on tasks/ship.rake:
```diff
@@ -585,8 +580,16 @@ namespace :pl do
+
       # If we just shipped a tagged version, we want to make it immutable
-      files = Dir.glob("#{local_dir}/**/*").select { |f| File.file?(f) }.map do |file|
+      files = Dir.glob("#{local_dir}/**/*").select { |f| File.file?(f) and !f.include? "yaml" }.map do |file|
```
should we make `!f.include? "yaml"` more specific to our yaml file? EZbake also produces a yaml file, which we probably don't want to overwrite all the time

mwaggett replied:
oh yeah, I was gonna do that but forgot. Will do now.

melissa commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,43 @@ def config(args = { :target => nil, :format => :hash })
+      # This is to be consumed by beaker and later replaced with our metadata service.
+      #
+      def platform_data
+        if self.project && self.ref
```
Is this just checking to make sure `project` and `ref` are set? Those should be safe assumptions. Did you run into issues where they weren't set

mwaggett replied:
yes

melissa replied:
Yes to the first question or yes to the second question where I forgot the question mark? If the second, can you elaborate on the issues you ran into?  Again, I believe these should be safe assumptions to make. All things we ship should have a project and a ref.

mwaggett replied:
They have a project and a ref eventually, but there's a point where `config_to_yaml` is called (which calls `config_to_hash`, which calls `platform_data`) before those have been determined.

melissa replied:
huh, strange

melissa commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,43 @@ def config(args = { :target => nil, :format => :hash })
+          artifacts, _ = Pkg::Util::Net.remote_ssh_cmd(self.builds_server, cmd, true)
+          artifacts = artifacts.split("\n")
+          data = {}
+          Pkg::Util::Platform.platform_tags.each do |tag|
```
what happens if we've shipped packages for platforms that we don't have entries for in `Pkg::Util::Platform.platform_tags`? This is not unheard of, where we have vanagon building packages before the packaging repo knows about them.

melissa replied:
Could it make more sense to iterate over the files that you find that aren't in `repo_configs` and interpolate the platform tag from there?

mwaggett replied:
They won't get included. To me, this seems like a better option than the opposite - grabbing the artifacts and then trying to figure out what platforms they're for. I think we just need to be better about adding platforms in all the places until we get the metadata service going.

mwaggett replied:
I think it would involve a lot more grossness to figure out the platform tags from the artifact. especially because the way I'm grabbing artifacts is just grabbing all the files and there are somethings in there that don't map to a platform tag, like the SRPMS for example.

underscorgan commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,43 @@ def config(args = { :target => nil, :format => :hash })
+            end
+            data[tag] = { :artifact => artifact,
+                          :repo_config => repo_config,
+                        }
```
I feel like it might be good to only add the tag if at least one of artifact/repo_config is set, otherwise there are a bunch of empty tags. I think it's reasonable for anything consuming this to handle missing tags.

mwaggett replied:
I feel like it's equally reasonable for anything consuming this to skip over platforms with missing artifacts/repo_configs. I can change that if you think it's better, just not sure what's best.

underscorgan commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,43 @@ def config(args = { :target => nil, :format => :hash })
+          data = {}
+          Pkg::Util::Platform.platform_tags.each do |tag|
+            _, _, arch = Pkg::Util::Platform.parse_platform_tag(tag)
+            artifact = artifacts.find { |e| e.include? Pkg::Util::Platform.artifacts_path(tag) and e.include? arch }
```
This isn't working for noarch debian packages. For rpms, you end up with the arch in the path even for noarch packages (like, http://builds.delivery.puppetlabs.net/puppetserver/5.0.0.master.SNAPSHOT.2017.05.04T1636/artifacts/el/7/PC1/x86_64/). But, the actual artifact for noarch deb things is something like `http://builds.delivery.puppetlabs.net/puppetserver/5.0.0.master.SNAPSHOT.2017.05.04T1636/artifacts/deb/xenial/PC1/puppetserver_5.0.0.master-0.1SNAPSHOT.2017.05.04T1636puppetlabs1_all.deb` which isn't going to match the arch. It was picking up the `.changes` file in the server deb which isn't what you want

melissa replied:
Debs don't always include arch. The path to the package doesn't include the architecture, and the package name only includes the architecture when it's an arch dependent package. Otherwise, it just has `all` in place of the architecture. For example, puppetdb: `http://builds.puppetlabs.lan/puppetdb/4.4.0/artifacts/deb/trusty/PC1/puppetdb_4.4.0-1puppetlabs1_all.deb`

mwaggett replied:
ah you're right, will fix

mwaggett replied:
ahhh will fix

underscorgan commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,43 @@ def config(args = { :target => nil, :format => :hash })
+          data = {}
+          Pkg::Util::Platform.platform_tags.each do |tag|
+            _, _, arch = Pkg::Util::Platform.parse_platform_tag(tag)
+            artifact = artifacts.find { |e| e.include? Pkg::Util::Platform.artifacts_path(tag) and (e.include?("all") || e.include?(arch)) }
```
Have you run with this? I'm curious what's going to happen for deb noarches where there's a .changes for i386 but you really want the 'all' .deb for both x86_64 and i386

mwaggett replied:
@underscorgan I'm building right now, so we'll see. I'm hoping it checks the 'all' first and just goes with that.

mwaggett replied:
oh wait, but it's gonna go through files one by one and if it checks the .changes file first, it'll grab that.. hmm. I'll keep thinking.

mwaggett replied:
I did work though..

underscorgan replied:
that feels fragile.

mwaggett replied:
I know, gonna change it.

underscorgan commented:
@mwaggett do you have p-a and puppetserver output with the changes? I like looking at the files :)

mwaggett commented:
@underscorgan Working on it. Still not quite cleaned up.

underscorgan commented:
cool cool

mwaggett commented:
updated puppet-agent yaml: http://builds.delivery.puppetlabs.net/puppet-agent/ac700b4d7e70783701e13079e0c7f3a635f3fe9a/artifacts/

updated puppetserver yaml: http://builds.delivery.puppetlabs.net/puppetserver/5.0.0.master.SNAPSHOT.2017.05.05T1637/artifacts/

melissa commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,42 @@ def config(args = { :target => nil, :format => :hash })
+          artifacts, _ = Pkg::Util::Net.remote_ssh_cmd(self.builds_server, cmd, true)
+          artifacts = artifacts.split("\n")
+          data = {}
+          Pkg::Util::Platform.platform_tags.each do |tag|
```
Since we're doing this, can you open a PR to winston to update the hash in this repo with the new platform? Or at least make sure that task is there? (I can't remember if it is or not)

mwaggett replied:
@melissa It's in there:
https://github.com/puppetlabs/winston/blob/master/tasks/platform-addition.rake#L156
https://github.com/puppetlabs/winston/blob/master/tasks/platform-removal.rake#L76

melissa replied:
saweet!

melissa commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,42 @@ def config(args = { :target => nil, :format => :hash })
+            if Pkg::Util::Platform.get_attribute(tag, :package_format) == 'deb'
+              artifact = artifacts.find { |e| e.include? Pkg::Util::Platform.artifacts_path(tag) and (e.include?("all") || e.include?("#{arch}.deb")) }
+              repo_config = "./repo_configs/deb/pl-#{self.project}-#{self.ref}-#{Pkg::Util::Platform.get_attribute(tag, :codename)}.list" if artifact
+            else
```
Make this an elseif rpm, then make the else error out with a good error message so we know the platform hash needs updating or something

mwaggett replied:
I can do something like that, but this needs to include rpms, dmgs, msis, etc. as well.

melissa replied:
The point still stands. We should bail with a reasonable error message when we're dealing with a package type we don't know how to deal with. I'd also suggest using a case statement.

Speaking of different package types, is there something we can do to check if there are any unreported packages? This wouldn't be a blocker, it would just be a nice to have, maybe in the next iteration

mwaggett replied:
Do you mean artifacts we found but didn't associate with a platform? Right now, I'm just looking at all files in a subdirectory under artifacts, which includes things like the .changes file. so I think that might be a bit of work and idk if it's worth it right now.

mwaggett replied:
@melissa updated

melissa commented on lib/packaging/config.rb:
```diff
@@ -69,13 +69,42 @@ def config(args = { :target => nil, :format => :hash })
+              artifact = artifacts.find { |e| e.include? Pkg::Util::Platform.artifacts_path(tag) and (e.include?("all") || e.include?("#{arch}.deb")) }
+              repo_config = "./repo_configs/deb/pl-#{self.project}-#{self.ref}-#{Pkg::Util::Platform.get_attribute(tag, :codename)}.list" if artifact
+            else
+              artifact = artifacts.find { |e| e.include? Pkg::Util::Platform.artifacts_path(tag) and e.include?(arch) }
```
Where does beaker want this path to start? I was under the impression it should start in the directory where the yaml file lives, (so, the artifacts directory). Then the artifacts path would be `el/7/...` and the repo config path would be `../repo_configs/el/7/...`. I see they both are now `./artifacts` and `./repo_configs`. Did you okay that with the beaker team?

mwaggett replied:
I didn't, but this was easier for me and doesn't seem like it would be especially difficult for beaker to just know to start in the <project>/<ref> directory.

melissa replied:
Make sure they're on the same page

mwaggett replied:
They're cool with this.
